### PR TITLE
feat: add ShutdownWithContext

### PR DESCRIPTION
**Pull request by li-jin-gou**
pr for https://github.com/valyala/fasthttp/issues/1389
- add `ShutdownWithContext` function to support shutdown with timeout and timeout happen  after all listener already close.

### CODE REVIEW

File: server.go [hunk 1]
```diff
@@ -1856,6 +1856,57 @@ func (s *Server) Shutdown() error {
...
+		if open := atomic.LoadInt32(&s.open); open == 0 {
+			break
+		}
+		// This is not an optimal solution but using a sync.WaitGroup
+		// here causes data races as it's hard to prevent Add() to be called
+		// while Wait() is waiting.
+		select {
+		case <-ctx.Done():
+			err = ctx.Err()
+			break END
```
Comment (erikdubbelboer) [inline server.go:1899]: I'm wondering, should this then force close all connections? What does `net/http` do in this case?

**Comment by pjebs** [comment]
`func (s *Server) Shutdown() error {` should be equivalent to `s.ShutdownWithContext (context.Background())`

File: server.go (hunk 1 above)
Comment (pjebs) [reply server.go:1899]: https://cs.opensource.google/go/go/+/refs/tags/go1.19.2:src/net/http/server.go;l=2731

Comment (li-jin-gou) [reply server.go:1899]: > I'm wondering, should this then force close all connections? What does `net/http` do in this case?

The use of `shutdown` does not force all connections to close, but rather uses context to control the timeout after all `listener` have been closed.

File: server_test.go [hunk 2]
```diff
@@ -3597,6 +3597,50 @@ func TestShutdownCloseIdleConns(t *testing.T) {
 	}
 }
 
+func TestShutdownWithContext(t *testing.T) {
+	t.Parallel()
+
+	ln := fasthttputil.NewInmemoryListener()
+	s := &Server{
+		Handler: func(ctx *RequestCtx) {
+			ctx.Success("aaa/bbb", []byte("real response"))
```
Comment (erikdubbelboer) [inline server_test.go:3606]: Can you add a `time.Sleep(time.Millisecond*100)` here and test that `ShutdownWithContext` returns correctly after the timeout? The test you have now just tests the code that is similar to what `TestShutdown` already tests. Thanks!

**Review by erikdubbelboer — CHANGES_REQUESTED** [review]

File: server_test.go (hunk 2 above)
Comment (li-jin-gou) [reply server_test.go:3606]: thanks

**Comment by erikdubbelboer** [comment]
Thanks!

**Comment by pjebs** [comment]
When will it be tagged?

**Comment by erikdubbelboer** [comment]
Next week I'll tag a release.

**Review by skydiver1980 — APPROVED** [review]

**Review by skydiver1980 — APPROVED** [review]

**Pull request merged**